<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Кристина, привет!👋</b> 

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 
 

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

     
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



# Анализ данных о продажах игр

- Автор:Вавилина Кристина
- Дата:

### Цели и задачи проекта

Основная цель проекта — исследовать данные о продажах игр, выявить основные тенденции и подготовить их для дальнейшего анализа. Для этого выполнены следующие задачи:

Загрузка и предварительное изучение данных.

Очистка данных от пропусков и дубликатов.

Приведение данных к единому формату.

Фильтрация и категоризация данных.


### Описание данных

Датасет содержит информацию о видеоиграх, включая название, платформу, год выпуска, жанр, продажи в разных регионах, оценки пользователей и критиков, а также рейтинг.

 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64 (изменить на int)
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object (изменить на float)
 6   JP sales         16956 non-null  object (изменить на float)
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object (изменить на float)
 10  Rating           10085 non-null  object 

### Содержимое проекта

Загрузка данных и знакомство с ними.

Проверка ошибок и предобработка данных.

Фильтрация данных по году выпуска.

Категоризация данных по оценкам пользователей и критиков.

Итоговый анализ и выводы.

---

 
<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Отлично по описанию. Цели и задачи ясны. Контекст проекта будет понятен любому потенциальному читателю.



## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd

df = pd.read_csv('/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [3]:
x = df.shape[0]

In [4]:
df.shape[1]

11

In [5]:
df.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

**Провели начальный анализ:** 
выявлено наличие пропусков в нескольких столбцах и некорректных типов данных.

Итоговые размеры данных:

Количество строк: 16956.

Количество столбцов: 11.

**Вывод:**

Есть пропуски в столбцах critic_score, user_score и rating.

Некоторые столбцы, например year_of_release, имеют неверный тип данных.

 
<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Все верно по проблемам в данных.
    
Единственное, можно сразу перечислить все стролбцы, где ожидаем увидеить численный тип, но не видим.



<div class="alert alert-info">
В описание данных добавила информацию о будущих изменениях типа данных в столбцах
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Cнова привет. Спасибо за заметное оформление твоих комментариев. По исправлениям отлично. По типам данных все увидел.
</div> 


---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [6]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
columns_snake_case = ['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']

In [8]:
df.columns = columns_snake_case

In [9]:
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [10]:
# Заполняем пропуски года выпуска 0 и меняем тип данных
df['year_of_release'] = df['year_of_release'].fillna(0)
df['year_of_release'] = df['year_of_release'].astype('int64')

In [11]:
# Меняем тип данных с помощью цикла
for column in ['na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score']:
    df[column] = pd.to_numeric(df[column], errors='coerce')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍  Хорошее решение
</div>   

Названия столбцов приведены к стилю snake_case

Изменён тип данных в столбцах:

year_of_release преобразован в int64.

Числовые данные в столбцах user_score и critic_score обработаны с заменой некорректных значений на NaN.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [12]:
df.isna().sum()

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [13]:
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     0.000000
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

In [14]:
def mean_group_cr_score(row):
    if pd.isna(row['eu_sales']):
        group = df[(df['platform'] == row['platform']) & 
                   (df['year_of_release'] == row['year_of_release'])]
        return group['eu_sales'].mean()
    else:
        return row['eu_sales']

df['eu_sales'] = df.apply(mean_group_cr_score, axis=1)

In [15]:
def mean_group_us_score(row):
    if pd.isna(row['jp_sales']):
        group = df[(df['platform'] == row['platform']) & 
                   (df['year_of_release'] == row['year_of_release'])]
        return group['jp_sales'].mean()
    else:
        return row['jp_sales']

df['jp_sales'] = df.apply(mean_group_us_score, axis=1)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 По заполнению пропусков здесь удобно также использовать функцию .transform() после .groupby()


```
#Заполнение продаж средним значением
avg_eu_sales = df.groupby(['year_of_release', 'platform'])['eu_sales'].transform('mean')
df['eu_sales'] = df['eu_sales'].fillna(avg_eu_sales)

avg_jp_sales = df.groupby(['year_of_release', 'platform'])['jp_sales'].transform('mean')
df['jp_sales'] = df['jp_sales'].fillna(avg_jp_sales)
```

Подробнее можно почитать <a href="https://dfedorov.spb.ru/pandas/%D0%9F%D0%BE%D0%BD%D0%B8%D0%BC%D0%B0%D0%BD%D0%B8%D0%B5%20%D1%84%D1%83%D0%BD%D0%BA%D1%86%D0%B8%D0%B8%20transform%20%D0%B2%20Pandas.html">здесь</a>




<div class="alert alert-info">
Большое спасибо за совет, правда, эта функция очень удобна и делает код меньше
</div>

<div class="alert alert-block alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2><s> Комментарий ревьюера #2 <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Эх, а мы перепутали. Заполнение группировкой у нас логично сделать для региональных прожаж (там и пропусков мало, и хоть какая-то логика прослеживается), а вот по оценкам вряд ли корректно сможем восстановить. Оценки очень индивидуальные для каждой игры. Оценки мы можем заполнять исключительно заглушкой «-1». 
    
Прошу прощения, не обратил на это внимание при первой проверке.

</div>

<div class="alert alert-info">
Заполнение группировкой теперь по региональным продажам, ниже оценки заполнила заглушкой
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Все верно, спасибо!
</div>

In [16]:
df = df.dropna(subset=['name'])

In [17]:
df.isna().sum()

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8712
user_score         9266
rating             6869
dtype: int64

In [18]:
df['critic_score'] = df['critic_score'].fillna(-1)

In [19]:
df['user_score'] = df['user_score'].fillna(-1)

In [20]:
df['rating'] = df['rating'].fillna("неизвестно")

In [21]:
df.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

Выявлены пропуски в столбцах:

critic_score: 51.4%.

user_score: 54.7%.

rating: 40.5%.

Выполнена обработка:

Пропуски в числовых столбцах заменены средними значениями по группе платформы и года выпуска.

Строки с критическими пропусками удалены.

Особенно заметна проблема в столбцах с оценками, где пропуски могли возникнуть из-за отсутствия отзывов от критиков или пользователей на момент сбора данных. Эти особенности могли быть характерны для игр, выпущенных в ранних годах или на менее популярных платформах.Возможно в специфических жанрах.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Ага, согласен по гипотезе.




<div class="alert alert-block alert-danger">
<h2><s> Комментарий ревьюера <a class="tocSkip"></h2>
  
<b>На доработку🤔:</b>
 
Важно и эти пропуски обработать.
    
В категориальной переменной rating можем прямо заменить пропуски на категорию «неизвестно», тем более, что предполагаем, что на месте пропусков она и подразумевается.
    
А для количественных переменных подойдет замена пропусков на нереалистичное значение, по которому их можно будет удобно отфильтровать в будущем. Например, таким значением может быть отрицательное число «-1».
</div>


<div class="alert alert-info">
Все пропуски обработаны:
    
- в переменной rating пропуски заменены на категорию «неизвестно»
    
- в количественных переменных 'eu_sales', 'jp_sales', 'critic_score', 'user_score' пропуски заменены на значение "-1"
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Ок. Нужно только поменять, к каким столбцам что применяем теперь :) 
</div> 


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [22]:
unique_combinations_count = df[['name', 'platform', 'year_of_release', 'genre', 'rating']].nunique()

unique_combinations_count

name               11559
platform              31
year_of_release       38
genre                 24
rating                 9
dtype: int64

In [23]:
# Приводим название и жанр к нижнему регистру
df['name'] = df['name'].str.lower()

df['genre'] = df['genre'].str.lower()

# Приводим рейтинг к верхнему регистру
df['rating'] = df['rating'].str.upper()

In [24]:
unique_combinations_count = df[['name', 'platform', 'year_of_release', 'genre', 'rating']].nunique()

unique_combinations_count

name               11559
platform              31
year_of_release       38
genre                 12
rating                 9
dtype: int64

In [25]:
df.duplicated().sum()

241

In [26]:
# Сохраняем количество строк до удаления дубликатов
initial_row_count = df.shape[0]

# Удаляем дубликаты
df_no_duplicates = df.drop_duplicates()

# Сохраняем количество строк после удаления дубликатов
final_row_count = df_no_duplicates.shape[0]

initial_row_count - final_row_count

241

In [27]:
# сколько всего строк удалили с момента загрузки датафрейма
x - final_row_count

243

In [28]:
# какую долю строк удалили с момента загрузки датафрейма
round(final_row_count/ x, 2)

0.99

Выявлено 241 явный дубликат, которые были удалены.

После преобразования данных в нижний регистр, число уникальных значений в жанрах игры изменилось с 24 на 12

Данные готовы к дальнейшему анализу

 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 
    
Выше нужно доработать по пропускам, а здесь − пересчитать.

<div class="alert alert-info">
Перепроверила данные после изменений, они остались прежними 
</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера  #2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

По проекту важно подсчитать финально, какую долю данных мы удалили по всем-всем преобразованиям после загрузки датасета. Я бы вначале сохранил количество строк в переменную, а здесь сравнил бы с тем, что получилось. У нас данные удалялись не только из-за дублей. 
       
</div>


<div class="alert alert-info">
После всех преобразований мы удалили 243 строки, доля сотавила 0.99
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера #2 <a class="tocSkip"> </h2>

👍 Ок.
</div>

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [29]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] < 2014)].copy()

In [30]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12980 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12980 non-null  object 
 1   platform         12980 non-null  object 
 2   year_of_release  12980 non-null  int64  
 3   genre            12980 non-null  object 
 4   na_sales         12980 non-null  float64
 5   eu_sales         12980 non-null  float64
 6   jp_sales         12980 non-null  float64
 7   other_sales      12980 non-null  float64
 8   critic_score     12980 non-null  float64
 9   user_score       12980 non-null  float64
 10  rating           12980 non-null  object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.2+ MB


После фильтрации датафрейм содержит 12980 строк

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍  Точно 
</div>  
       


---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [31]:
df_actual['user_score_group'] = pd.cut(df_actual['user_score'], bins=[0, 3, 8, 11], labels=["низкая оценка", "средняя оценка", "высокая оценка"],right=False)

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Верно по `right=False`
</div>  


In [32]:
count_user_group = df_actual.groupby('user_score_group')['name'].count()

count_user_group

user_score_group
низкая оценка      117
средняя оценка    4148
высокая оценка    2307
Name: name, dtype: int64

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [33]:
df_actual['critic_score_group'] = pd.cut(df_actual['critic_score'], bins=[0, 30, 80, 101], labels=["низкая оценка", "средняя оценка", "высокая оценка"],right=False)

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 
</div>  


In [34]:
count_critic_group = df_actual.groupby('critic_score_group')['name'].count()

count_critic_group

critic_score_group
низкая оценка       55
средняя оценка    5500
высокая оценка    1712
Name: name, dtype: int64

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [35]:
count_game_group = df_actual.groupby(['user_score_group','critic_score_group'])['name'].count()

count_game_group

user_score_group  critic_score_group
низкая оценка     низкая оценка           17
                  средняя оценка          78
                  высокая оценка           1
средняя оценка    низкая оценка           30
                  средняя оценка        3210
                  высокая оценка         649
высокая оценка    низкая оценка            1
                  средняя оценка        1175
                  высокая оценка        1028
Name: name, dtype: int64

Результат категоризации:

Игры с высокой оценкой критиков: 1781

Игры с высокой оценкой пользователей: 2381

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [36]:
platform_group = df_actual.groupby('platform')['name'].count()
platform_group.sort_values(ascending=False).head(7)


platform
PS2     2154
DS      2146
Wii     1294
PSP     1199
X360    1138
PS3     1107
GBA      826
Name: name, dtype: int64

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Здесь все корректно.
</div>  
	  


---

## 5. Итоговый вывод

 Мы провели анализ данных о продажах видеоигр, выполнены следующие шаги:

#### Предварительная обработка данных:

- Приведение столбцов к единому формату (snake_case).
- Исправление некорректных типов данных (например, преобразование года выпуска в целое число).
- Обработка пропусков: пропущенные значения в числовых столбцах заменены средними значениями по платформе и году выпуска, строки с критическими пропусками удалены.
-Удалены 241 явный дубликат, а уникальные значения в категориальных данных нормализованы.

#### Фильтрация данных:

Данные ограничены периодом с 2000 по 2013 год. 

После фильтрации получен новый датафрейм df_actual с 12,980 строками.

#### Категоризация данных:

- Игры разделены по оценкам пользователей: высокая (от 8 до 10 включительно), средняя (от 3 до 8) и низкая (от 0 до 3).
- Игры разделены по оценкам критиков: высокая (от 80 до 100 включительно), средняя (от 30 до 80), низкая (от 0 до 30).
- Проведено сравнение пересечений категорий пользователей и критиков.

#### Выявлены основные платформы:

Топ-7 платформ по количеству выпущенных игр: PS2, DS, Wii, PSP, X360, PS3, GBA.

#### Срез данных
Период: 2000–2013 годы.
Количество строк после фильтрации: 12,980.

Категориальные поля:
- user_score_group: категория оценки пользователей.
- critic_score_group: категория оценки критиков.

#### Новые поля
В исходный датасет добавлены:

- user_score_group — категория оценки пользователей.
- critic_score_group — категория оценки критиков.

#### Основной вывод
Данные успешно подготовлены и очищены для дальнейшего анализа. В результате категоризации были выделены основные характеристики игр, включая платформы, высоко оцененные пользователями и критиками. Эти результаты могут использоваться для выявления трендов, построения прогнозов продаж и анализа предпочтений игроков.

<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Замечательно по содержанию и оформлению. 
</div>  
	  


<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**
    
Спасибо за твой проект! Внимательная работа, практически все решения верные и оптимальные, но кое-что нужно доработать и кое-что можно улучшить, чтобы проект стал совсем замечательным.
    
**Отмечу положительные моменты проекта🚀**
    
* Отличное оформление. Есть описание, комментарии, выводы.
* Хорошее владение кодом, инструментами предобработки. Все сложные моменты выполнены отлично.
* Хорошая аргументация решений. Во всех случаях понятно, почему принимали те или иные решения.
    
**На что стоит обратить внимание🤔**
    
* Важно доработать по абсолютно всем пропускам в проекте. У нас нет супер-неопределенных ситуаций, в каждой можем применить преобразование, которое улучшит данные и позволит дальше работать с ними.
* Эти и остальные замечания я также написал по ходу проекта.
    
**Желаю удачи и жду твой проект на повторное ревью!**😉
    
    
</div>




<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий ко второй проверке**

Спасибо за качественную доработку. Практически все верно. Однако еще кое-что осталось.

* Важно поменять подходы к обработке пропусков: региональные продажи мы можем восстановить группировкой по году и платформе, а вот рейтинги придется заменять на заглушку типа «-1». 
    
* Также будет правильнее рассчитать финальную долю данных, которые мы удалили. Туда будут входить не только дубликаты, но и все-все, что мы удаляли с момента загрузки датасета.

**Желаю удачи и жду твой проект на повторное ревью!**😉
</div>


<div style="border:solid Chocolate 2px; padding: 40px">

**Комментарий к финальной проверке**

Спасибо за доработку. Критичных комментариев нет. 

Если что-то пропустил или остались вопросы, напиши в телеграм @kriill

Желаю удачи в дальнейшей учебе!

</div>
